src/celeritas/em/msc/detail/MscStepToGeo.hh

```diff
@@ -0,0 +1,189 @@
+    };
+
+  public:
+    // Construct from MSC data
+    inline CELER_FUNCTION MscStepToGeo(UrbanMscRef const& shared,
+                                       UrbanMscHelper const& helper,
+                                       Energy energy,
+                                       real_type lambda,
+                                       real_type range);
+
+    // Calculate the geometrical step
+    inline CELER_FUNCTION result_type operator()(real_type tstep) const;
+
+  private:
+    UrbanMscRef const& shared_;
+    UrbanMscHelper const& helper_;
+    real_type energy_;
+    real_type lambda_;
+    real_type range_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from MSC data.
+ */
+CELER_FUNCTION MscStepToGeo::MscStepToGeo(UrbanMscRef const& shared,
+                                          UrbanMscHelper const& helper,
+                                          Energy energy,
+                                          real_type lambda,
+                                          real_type range)
+    : shared_(shared)
+    , helper_(helper)
+    , energy_(energy.value())
+    , lambda_(lambda)
+    , range_(range)
+{
+    CELER_EXPECT(energy_ > 0);
+    CELER_EXPECT(lambda_ > 0);
+    CELER_EXPECT(range_ > 0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the geometry step length for a given true step length.
+ */
+CELER_FUNCTION auto MscStepToGeo::operator()(real_type tstep) const
+    -> result_type
+{
+    result_type result;
+    result.alpha = MscStep::small_step_alpha();
+    if (tstep < shared_.params.min_step())
+    {
+        // Geometrical path length = true path length for a very small step
+        result.step = tstep;
+    }
+    else if (tstep <= lambda_ * shared_.params.tau_small)
+    {
+        // Very small distance to collision (less than tau_small paths)
+        result.step = min(tstep, lambda_);
+    }
+    else if (tstep < range_ * shared_.params.dtrl())
+    {
+        // Small enough distance to assume cross section is constant
+        // over the step: z = lambda * (1 - exp(-tau))
+        real_type tau = tstep / lambda_;
+        result.step = (tau < shared_.params.tau_limit)
+                          ? tstep * (1 - tau / 2)
+                          : lambda_ * (1 - std::exp(-tau));
+    }
+    else
+    {
+        // Eq 8.8: mfp_slope = (lambda_f / lambda_i) = (1 - alpha * tstep)
+        real_type mfp_slope;
+        if (energy_ < value_as<Mass>(shared_.electron_mass) || tstep == range_)
+        {
+            // Low energy or range-limited step
+            // (For range-limited step, the final cross section and thus MFP
+            // slope are zero).
+            result.alpha = 1 / range_;
+            mfp_slope = 1 - result.alpha * tstep;
```

Comment:
I hit the assertion failure in `fastpow()` when the step was range-limited (`tstep == range_`) and the calculated `mfp_slope` was not exactly zero (e.g. -4.89081360221648e-17).

Reply:
Aha, thanks. I've fixed it with a `max`. Guess I forgot that with floating point arithmetic, `1 - (1/x) * x` isn't necessarily zero 😬